Addresses #19 - Fix and update scraping and data dumping

satyamtg opened the pull request:
Addresses #19 by scraping topicwise. The scraping logic has been updated and following new arguments are introduced -
- --topics - Comma-seperated list of all categories (spelling should be same as in the list on ted.com/talks
- --max-videos-per-topic - Maximum number of videos to scrape in each topic
- --output - The output folder

The scraper now uses a supplied folder as the output folder and directly builds files into the build directory inside it. This is a similar structure as the youtube scraper. Previously it had 3 directories inside the output folder.

Since the scraper now scrapes with a given list of topics, it now dumps 2 json files, one containing the video data and the other containing the list of topics.

Status - The scraping logic and json dumping work now. It can be tested as follows -
```
ted2zim --topics="augmented reality" --max-videos-per-topic=55 --debug --output="output"
```
The edge case of passing wrong topic names would be handled later. Currently, it would just dump the supplied topics in the category data json file.

It would be very easy to add language filtering in this scraping logic. It will be addressed soon.

rgaudin commented on ted2zim/entrypoint.py:
```diff
@@ -14,11 +14,29 @@ def main():
     )
 
+    parser.add_argument(
+        "--categories",
```
As those are topics, can we use `--topics` instead?

satyamtg replied:
Yup. Renamed categories to topics

rgaudin commented on ted2zim/entrypoint.py:
```diff
@@ -14,11 +14,29 @@ def main():
+    )
+
+    parser.add_argument(
+        "--max-videos",
```
`--max-videos-per-topic` would make it clearer

satyamtg replied:
Done. Renamed to --max-videos-per-topic

rgaudin commented on ted2zim/entrypoint.py:
```diff
@@ -14,11 +14,29 @@ def main():
+
+    parser.add_argument(
+        "--max-videos",
+        help="Max number of videos to scrape in each topic. Pass 'max' if you want to scrape all",
```
Scraping all should be the default behavior. If you want to postpone that, we can open a ticket but then the special value should be `all` rather than `max`.

satyamtg replied:
We pass a large number as default value (for now). This makes scraping all as the default behaviour.

rgaudin commented on ted2zim/scraper.py:
```diff
@@ -1,129 +1,145 @@
+
+        if max_videos.isdigit():
+            self.max_videos = int(max_videos)
+        elif max_videos == "max":
```
this will crash on non digit value other than max. We don't want to crash on `__init__()`.
init should just set the variables and the handling go to `run()`.
Crashes on user inputs should be taken care of and a proper message returned (inside `run()`)

satyamtg replied:
Doesn't hold anymore. This has been removed now.

rgaudin commented on ted2zim/scraper.py:
```diff
@@ -1,129 +1,145 @@
+            video_allowance = None
+            page = 1
+            tot_videos_scraped = 0
+            if self.max_videos == "max":
```
this value should be for end-users only. If that's what the user wants, then set this value earlier.

satyamtg replied:
Its simply `video_allowance = self.max_videos_per_topic` now.

rgaudin commented on ted2zim/scraper.py:
```diff
@@ -199,53 +217,58 @@ def extract_video_info(self, url):
+        logger.debug(
+            f"Dumping {len(self.videos)} videos into {self.ted_videos_json} and {len(self.categories)} categories into {self.ted_categories_json}"
+        )
+        video_data = json.dumps(self.videos, indent=4, separators=(",", ": "))
```
remove the separators arg. those are the defaults in py3.

satyamtg replied:
Done

rgaudin commented on ted2zim/utils.py:
```diff
@@ -1,41 +1,23 @@
-__license__ = "GPLv3"
-
 
 def create_absolute_link(base, rel_url):
```
I think is is not needed. It's used once and does very little

satyamtg replied:
Yup. You're correct. Removed this.

satyamtg commented:
> I've added a couple of changes that you can do right now but it's mostly OK.
> Looking at the code through the PR, I'm seeing a lot of room for improvement so I'll open some small issues. We'll tackle them all once we're on master.

@rgaudin thanks for these suggestions. I've addressed them. There's surely much room for improvement and we'll surely do it once we have a working scraper on master.

rgaudin commented on ted2zim/scraper.py:
```diff
@@ -700,7 +687,7 @@ def create_zim(static_folder, zim_path, title, description, name):
         "scraper": "Ted scraper : https://github.com/openzim/ted",
         "source": "https://new.ted.com",
-        "tags": "_category:ted;ted",
+        "tags": "_topic:ted;ted",
```
here this is different. `_category` is a special tag that we use to categorize ZIM files (see the list of folders on https://download.kiwix.org/zim/

satyamtg replied:
Thanks for pointing that out
